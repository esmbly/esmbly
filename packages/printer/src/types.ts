export type Callback = (error: Error | null | undefined) => void;

export type Encoding = string | undefined;

export type WriteFn = (
  chunk: string,
  encoding?: Encoding,
  callback?: Callback,
) => boolean;
