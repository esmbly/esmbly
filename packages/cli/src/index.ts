import yargs, { Arguments } from 'yargs';
import * as init from './commands/init';
import * as run from './commands/run';

export function runCLI(): Arguments {
  return yargs
    .scriptName('esmbly')
    .command(init)
    .command(run)
    .help().argv;
}
