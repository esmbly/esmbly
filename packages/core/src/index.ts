import { File, RunConfig } from '@esmbly/types';
import { parse } from './parse';
import { validateConfig, validateInputFormat } from './validate';

export async function run(config: RunConfig): Promise<File[]> {
  validateConfig(config);

  let files: File[] = [];
  const trees = parse(config.input, config.transformers[0]);

  for (const transformer of config.transformers) {
    if (transformer.before) {
      await transformer.before();
    }

    validateInputFormat(trees, transformer);

    if (transformer.transform) {
      await transformer.transform(trees);
    }

    if (transformer.createFiles) {
      files = [
        ...files,
        ...(await transformer.createFiles(trees, config.output)),
      ];
    }

    if (transformer.after) {
      await transformer.after();
    }
  }

  return files;
}
