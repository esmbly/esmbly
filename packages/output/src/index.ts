type Callback = (error: Error | null | undefined) => void;
type Encoding = string | undefined;
type WriteFn = (
  chunk: string,
  encoding?: Encoding,
  callback?: Callback,
) => boolean;

class Output {
  private stdout: NodeJS.WriteStream | null;

  private stderr: NodeJS.WriteStream | null;

  public constructor() {
    this.stdout = process ? process.stdout : null;
    this.stderr = process ? process.stderr : null;
  }

  public get isTTY(): boolean {
    if (this.stdout) {
      return this.stdout.isTTY === true;
    }
    return false;
  }

  public err(chunk: string, encoding?: Encoding, callback?: Callback): boolean {
    if (this.stderr) {
      return this.stderr.write(chunk, encoding, callback);
    }
    return false;
  }

  public out(chunk: string, encoding?: Encoding, callback?: Callback): boolean {
    if (this.stdout) {
      return this.stdout.write(chunk, encoding, callback);
    }
    return false;
  }

  public clearLine(...args: WriteFn[]): void {
    if (this.isTTY) {
      args.forEach(fn => fn.call(this, '\x1b[999D\x1b[K'));
    }
  }

  public setErrorStream(
    stream: NodeJS.WriteStream | NodeJS.WritableStream | null,
  ): void {
    this.stderr = stream as NodeJS.WriteStream;
  }

  public setOutStream(
    stream: NodeJS.WriteStream | NodeJS.WritableStream | null,
  ): void {
    this.stdout = stream as NodeJS.WriteStream;
  }
}

export default new Output();
