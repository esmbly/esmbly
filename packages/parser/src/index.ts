import parse from './parse';

export default { parse };
