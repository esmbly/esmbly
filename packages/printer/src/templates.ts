import { File, Warning } from '@esmbly/types';
import chalk from 'chalk';
import { SourceLocation } from '@babel/types';

function location(file?: File, loc?: SourceLocation | null): string {
  if (!file) {
    return '';
  }

  const line = loc ? loc.start.line : null;
  const column = loc ? loc.start.column : null;
  const position = line !== null && column !== null ? `:${line}:${column}` : '';
  const path = `${file.dir}/${file.name}${file.type}${position}`.replace(
    /^\/+/g,
    '',
  );
  return `${chalk.dim('at:')} ${chalk.cyan(path)}`;
}

function removeLeadingWhitespace(
  str: string,
): {
  line: string;
  trimmed: number;
} {
  let line = '';
  let foundChar = false;
  let trimmed = 0;

  for (let i = 0; i < str.length; i += 1) {
    const isWhitespace = str[i] === ' ';

    if (isWhitespace && !foundChar) {
      trimmed += 1;
    } else {
      if (!isWhitespace) {
        foundChar = true;
      }

      line += str[i];
    }
  }

  return { line, trimmed };
}

function issue(file?: File, loc?: SourceLocation | null): string {
  if (!file || !loc) {
    return '';
  }

  const lines = file.content.toString().split('\n');
  const { line, trimmed } = removeLeadingWhitespace(lines[loc.start.line - 1]);
  const start = loc.start.column - trimmed;
  const end =
    loc.start.line === loc.end.line ? loc.end.column - trimmed : line.length;
  const leading = chalk.dim(line.slice(0, start));
  const highlighted = chalk.red(line.slice(start, end + 1));
  const trailing = chalk.dim(line.slice(end + 2));
  return `${leading}${highlighted}${trailing}`;
}

function url(issueUrl?: string): string {
  if (!issueUrl) {
    return '';
  }

  return `${chalk.dim('issue:')} ${chalk.cyan(issueUrl)}`;
}

export function warning({ info, loc, file, issueUrl }: Warning): string {
  return `
  ${chalk.yellow('warning')} ${chalk.dim(info)}

    ${issue(file, loc)}
  
  ${location(file, loc)}
  ${url(issueUrl)}

`;
}
