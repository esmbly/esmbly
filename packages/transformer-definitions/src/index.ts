import { File, Format, Output, SyntaxTree, Transformer } from '@esmbly/types';
import { createFiles } from './createFiles';

export function createTransformer(): Transformer {
  return {
    createFiles(trees: SyntaxTree[], output: Output[]): File[] {
      return createFiles(trees, output);
    },
    format: {
      files: [Format.Any],
      input: Format.TypeScript,
      output: Format.TypeScript,
    },
    name: 'Definitions',
    parserPlugins: ['typescript'],
  };
}
