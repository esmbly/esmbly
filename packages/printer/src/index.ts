import { Warning } from '@esmbly/types';
import chalk from 'chalk';
import { Callback, Encoding, WriteFn } from './types';
import * as templates from './templates';

class Printer {
  private stdout: NodeJS.WriteStream | null;
  private stderr: NodeJS.WriteStream | null;

  public constructor() {
    this.stdout = process ? process.stdout : null;
    this.stderr = process ? process.stderr : null;
  }

  public get isTTY(): boolean {
    if (this.stdout) {
      return this.stdout.isTTY === true;
    }

    return false;
  }

  public error(
    chunk: string,
    encoding?: Encoding,
    callback?: Callback,
  ): boolean {
    if (this.stderr) {
      return this.stderr.write(chunk, encoding, callback);
    }

    return false;
  }

  public print(
    chunk: string,
    encoding?: Encoding,
    callback?: Callback,
  ): boolean {
    if (this.stdout) {
      return this.stdout.write(chunk, encoding, callback);
    }

    return false;
  }

  public printWarnings(warnings: Warning[]): void {
    if (this.stdout) {
      warnings.forEach((warning: Warning) => {
        this.print(templates.warning(warning));
      });
    }
  }

  public forceEnableColors(): void {
    chalk.enabled = true;
  }

  public forceDisableColors(): void {
    chalk.enabled = false;
  }

  public clearLine(...args: WriteFn[]): void {
    if (this.isTTY) {
      args.forEach(fn => fn.call(this, '\x1b[999D\x1b[K'));
    }
  }

  public setErrorStream(
    stream: NodeJS.WriteStream | NodeJS.WritableStream | null,
  ): void {
    this.stderr = stream as NodeJS.WriteStream;
  }

  public setOutStream(
    stream: NodeJS.WriteStream | NodeJS.WritableStream | null,
  ): void {
    this.stdout = stream as NodeJS.WriteStream;
  }
}

export const printer = new Printer();
export { templates };
