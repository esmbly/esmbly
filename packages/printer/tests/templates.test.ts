import { FileType } from '@esmbly/types';
import * as templates from '../src/templates';

describe('templates', () => {
  it('provides a correct template for warnings', () => {
    expect(
      templates.warning({
        file: {
          content: 'function add(a: any): string {}',
          dir: 'src',
          name: 'add',
          type: FileType.JavaScript,
        },
        info: 'info about the warning',
        issueUrl: 'some url',
        loc: {
          end: { column: 20, line: 1 },
          start: { column: 14, line: 1 },
        },
      }),
    ).toMatchSnapshot();
  });
});
