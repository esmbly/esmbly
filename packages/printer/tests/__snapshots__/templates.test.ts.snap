// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`templates provides a correct template for warnings 1`] = `
"
  [33mwarning[39m [2minfo about the warning[22m

    [2mfunction add(a[22m[31m: any):[39m[2mstring {}[22m
  
  [2mat:[22m [36msrc/add.js:1:14[39m
  [2missue:[22m [36msome url[39m

"
`;
